src/Custom/Microsoft.Extensions.AI.OpenAIAdapter.cs

```diff
@@ -0,0 +1,1820 @@
+    /// <param name="response">The response to update.</param>
+    /// <param name="audioTranslation">The OpenAI audio translation.</param>
+    private static void UpdateResponseFromOpenAIAudioTranslation(SpeechToTextResponse response, AudioTranslation audioTranslation)
+    {
+        _ = Throw.IfNull(audioTranslation);
+
+        var segmentCount = audioTranslation.Segments.Count;
+
+        TimeSpan? endTime = null;
+        TimeSpan? startTime = null;
+        if (segmentCount > 0)
+        {
+            endTime = audioTranslation.Segments[segmentCount - 1].EndTime;
+            startTime = audioTranslation.Segments[0].StartTime;
+        }
+
+        // Update the response
+        response.RawRepresentation = audioTranslation;
+        response.Contents = [new TextContent(audioTranslation.Text)];
+        response.StartTime = startTime;
+        response.EndTime = endTime;
+        response.AdditionalProperties = new AdditionalPropertiesDictionary
+        {
+            [nameof(audioTranslation.Language)] = audioTranslation.Language,
+            [nameof(audioTranslation.Duration)] = audioTranslation.Duration
+        };
+    }
+
+    /// <summary>Converts an extensions options instance to an OpenAI options instance.</summary>
+    private static AudioTranslationOptions ToOpenAITranslationOptions(SpeechToTextOptions? options)
+    {
+        AudioTranslationOptions result = new();
+
+        if (options is not null)
+        {
+            if (options.AdditionalProperties is { Count: > 0 } additionalProperties)
+            {
+                if (additionalProperties.TryGetValue(nameof(result.Temperature), out float? temperature))
+                {
+                    result.Temperature = temperature;
+                }
+
+                if (additionalProperties.TryGetValue(nameof(result.ResponseFormat), out AudioTranslationFormat? responseFormat))
+                {
+                    result.ResponseFormat = responseFormat;
+                }
+
+                if (additionalProperties.TryGetValue(nameof(result.Prompt), out string? prompt))
+                {
+                    result.Prompt = prompt;
+                }
+            }
+        }
+
+        return result;
+    }
+}
+
+internal static class Throw
+{
+    /// <summary>
+    /// Throws an <see cref="System.ArgumentNullException"/> if the specified argument is <see langword="null"/>.
+    /// </summary>
+    /// <typeparam name="T">Argument type to be checked for <see langword="null"/>.</typeparam>
+    /// <param name="argument">Object to be checked for <see langword="null"/>.</param>
+    /// <param name="paramName">The name of the parameter being checked.</param>
+    /// <returns>The original value of <paramref name="argument"/>.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public static T IfNull<T>(T argument, string paramName = "")
```

Comment:
The paramName needs a [CallerArgumentExpression] attribute on it. Otherwise, every call site is going to have the wrong parameter name in it.